File: Backend/app/services/route_optimizer.py
"""
Route Optimizer Service
Uses Graph and NumPy for route optimization
"""

import numpy as np
from app.data_structures.graph import Graph


class RouteOptimizer:
    """
    Optimize travel routes using Graph + NumPy
    Demonstrates: graph algorithms, NumPy calculations
    """
    
    @staticmethod
    def calculate_distance(lat1, lon1, lat2, lon2):
        """
        Haversine formula using NumPy
        Demonstrates: NumPy trigonometric functions
        """
        R = 6371  # Earth radius in km
        
        # Convert to radians
        lat1, lon1, lat2, lon2 = np.radians([lat1, lon1, lat2, lon2])
        
        dlat = lat2 - lat1
        dlon = lon2 - lon1
        
        a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
        c = 2 * np.arcsin(np.sqrt(a))
        
        return R * c
    
    @staticmethod
    def optimize_route(cities):
        """
        Optimize route using greedy nearest neighbor
        Demonstrates: NumPy matrix operations, algorithms
        """
        if not cities or len(cities) <= 2:
            return {
                'cities': cities,
                'total_distance': 0,
                'route_order': list(range(len(cities)))
            }
        
        n = len(cities)
        
        # Create distance matrix using NumPy
        distances = np.zeros((n, n))
        
        for i in range(n):
            for j in range(n):
                if i != j:
                    distances[i][j] = RouteOptimizer.calculate_distance(
                        cities[i]['latitude'],
                        cities[i]['longitude'],
                        cities[j]['latitude'],
                        cities[j]['longitude']
                    )
        
        # Greedy nearest neighbor algorithm
        visited = np.zeros(n, dtype=bool)
        route = [0]
        visited[0] = True
        total_distance = 0
        
        for _ in range(n - 1):
            last = route[-1]
            
            # Find nearest unvisited city
            unvisited_distances = distances[last].copy()
            unvisited_distances[visited] = np.inf
            
            nearest = np.argmin(unvisited_distances)
            
            route.append(nearest)
            visited[nearest] = True
            total_distance += distances[last][nearest]
        
        # Reorder cities according to route
        optimized_cities = [cities[i] for i in route]
        
        return {
            'cities': optimized_cities,
            'total_distance': round(total_distance, 2),
            'route_order': route,
            'distance_matrix': distances.tolist()
        }

File: Frontend/js/route_optimizer.js
/**
 * Route Optimizer Module
 * Optimizes travel route using Nearest Neighbor algorithm
 */
const RouteOptimizer = {
    /**
     * Calculate distance between two points using Haversine formula
     * @param {number} lat1 
     * @param {number} lon1 
     * @param {number} lat2 
     * @param {number} lon2 
     * @returns {number} Distance in km
     */
    calculateDistance(lat1, lon1, lat2, lon2) {
        const R = 6371; // Radius of the earth in km
        const dLat = this.deg2rad(lat2 - lat1);
        const dLon = this.deg2rad(lon2 - lon1);
        const a =
            Math.sin(dLat / 2) * Math.sin(dLat / 2) +
            Math.cos(this.deg2rad(lat1)) * Math.cos(this.deg2rad(lat2)) *
            Math.sin(dLon / 2) * Math.sin(dLon / 2);
        const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        const d = R * c; // Distance in km
        return d;
    },

    deg2rad(deg) {
        return deg * (Math.PI / 180);
    },

    /**
     * Optimize route using Nearest Neighbor algorithm
     * @param {Array} cities - List of city objects with lat/lon
     * @returns {Object} Optimized route and total distance
     */
    optimizeRoute(cities) {
        if (!cities || cities.length <= 1) {
            return {
                route: cities,
                totalDistance: 0
            };
        }

        // Clone cities to avoid modifying original array
        let unvisited = [...cities];
        const route = [];
        let totalDistance = 0;

        // Start with the first city in the list (or user could select start)
        // For now, we assume the first selected city is the starting point
        let currentCity = unvisited.shift();
        route.push(currentCity);

        while (unvisited.length > 0) {
            let nearestCity = null;
            let minDistance = Infinity;
            let nearestIndex = -1;

            unvisited.forEach((city, index) => {
                // If lat/lon missing, assume 0 distance (or handle error)
                if (!currentCity.latitude || !city.latitude) return;

                const dist = this.calculateDistance(
                    currentCity.latitude, currentCity.longitude,
                    city.latitude, city.longitude
                );

                if (dist < minDistance) {
                    minDistance = dist;
                    nearestCity = city;
                    nearestIndex = index;
                }
            });

            if (nearestCity) {
                totalDistance += minDistance;
                currentCity = nearestCity;
                route.push(currentCity);
                unvisited.splice(nearestIndex, 1);
            } else {
                // If we can't calculate distance (missing coords), just add the rest
                route.push(...unvisited);
                break;
            }
        }

        return {
            route: route,
            totalDistance: Math.round(totalDistance)
        };
    }
};

// Expose to window
window.RouteOptimizer = RouteOptimizer;
